### PR TITLE
Explicit copyout changes

### PR DESCRIPTION
**Pull request by bsdjhb**
This is the rest of the changes to use explicit user caps.  After this I will start on the cleanups to remove *_c and retire the !EXPLICIT_USER_ACCESS code, etc.

If this is too big of a chunk I can split this up for review.

### CODE REVIEW

File: sys/cam/cam_periph.c [hunk 1]
```diff
@@ -932,7 +934,7 @@ cam_periph_mapmem(union ccb *ccb, struct cam_periph_map_info *mapinfo,
 	for (i = 0; i < numbufs; i++) {
 
 		/* Save the user's data address. */
-		mapinfo->orig[i] = *data_ptrs[i];
+		mapinfo->orig[i] = __USER_CAP(*data_ptrs[i], lengths[i]);
```
Comment (qwattash) [inline sys/cam/cam_periph.c:937]: I'm not super familiar with this part of the code, however it seems that `data_ptrs[]` holds `uint8_t*` and `struct ccb_dev_match` (for example) contains various pointers. I believe that I will have to do some `uint64_t` or `__ptr64` annotation for these to work in the purecap kernel. Do you know which userspace tool uses these ioctls?

File: sys/cam/ctl/ctl_frontend_ioctl.c [hunk 2]
```diff
@@ -367,7 +369,7 @@ ctl_ioctl_do_datamove(struct ctl_scsiio *ctsio)
 		ext_sglist = (struct ctl_sg_entry *)malloc(ext_sglen, M_CTL,
 							   M_WAITOK);
 		ext_sglist_malloced = 1;
-		if (copyin(ctsio->ext_data_ptr, ext_sglist, ext_sglen) != 0) {
+		if (copyin(ctsio->ext_data_uptr, ext_sglist, ext_sglen) != 0) {
```
Comment (qwattash) [inline sys/cam/ctl/ctl_frontend_ioctl.c:372]: It is unclear to me what sets `ext_data_uptr`, do we need a corresponding `__USER_CAP` conversion from it somewhere, or we assume that this will not work unless the userspace is purecap?

File: sys/cam/scsi/scsi_pass.c [hunk 3]
```diff
@@ -1238,7 +1242,9 @@ passcopysglist(struct cam_periph *periph, struct pass_io_req *io_req,
 				xpt_print(periph->path, "%s: copyout of %u "
 					  "bytes from %p to %p failed with "
 					  "error %d\n", __func__, len_to_copy,
-					  kern_ptr, user_ptr, error);
+					  kern_ptr,
+					  (void *)(__cheri_addr uintptr_t)user_ptr,
```
Comment (qwattash) [inline sys/cam/scsi/scsi_pass.c:1246]: Perhaps we should add a capability format specifier to printf, these casts are quite common.

File: sys/cam/scsi/scsi_pass.c [hunk 4]
```diff
@@ -1219,12 +1221,14 @@ passcopysglist(struct cam_periph *periph, struct pass_io_req *io_req,
...
 	     j < io_req->num_kern_segs;) {
-		uint8_t *user_ptr, *kern_ptr;
+		uint8_t * __capability user_ptr;
+		uint8_t *kern_ptr;
 
 		len_to_copy = min(user_sglist[i].ds_len -user_watermark,
 		    kern_sglist[j].ds_len - kern_watermark);
 
-		user_ptr = (uint8_t *)(uintptr_t)user_sglist[i].ds_addr;
+		user_ptr = __USER_CAP((uint8_t *)(uintptr_t)user_sglist[i].ds_addr,
```
Comment (qwattash) [inline sys/cam/scsi/scsi_pass.c:1230]: Now `__USER_CAP` accepts `uint64_t` without need to cast. I feel it is somewhat redundant as the macro already makes it explicit that the result will be a pointer.
As an aside (and as a note to self), going back to our last conversation about `bus_addr_t`, it seems that there are different interpretations of `bus_addr_t` across architectures on whether it holds a physical address, an opaque value of some kind or a virtual address. This seems to be one of those places that assumes it can hold a virtual address.

File: sys/compat/linuxkpi/common/src/linux_compat.c [hunk 5]
```diff
@@ -824,44 +826,44 @@ linux_remap_address(void **uaddr, size_t len)
...
 int
-linux_copyin(const void *uaddr, void *kaddr, size_t len)
+linux_copyin(const void * __capability uaddr, void *kaddr, size_t len)
 {
-	if (linux_remap_address(__DECONST(void **, &uaddr), len)) {
+	if (linux_remap_address(__DECONST(void * __capability *, &uaddr), len)) {
 		if (uaddr == NULL)
 			return (-EFAULT);
-		memcpy(kaddr, uaddr, len);
+		memcpy(kaddr, (void *)(__cheri_addr uintptr_t)uaddr, len);
```
Comment (qwattash) [inline sys/compat/linuxkpi/common/src/linux_compat.c:841]: This would fail in the purecap kernel. Should probably use `(__cheri_fromcap void *)uaddr` that keeps `uaddr` from becoming untagged in pruecap mode.

File: sys/compat/linuxkpi/common/src/linux_compat.c [hunk 6]
```diff
@@ -824,44 +826,44 @@ linux_remap_address(void **uaddr, size_t len)
...
 
 int
-linux_copyout(const void *kaddr, void *uaddr, size_t len)
+linux_copyout(const void *kaddr, void * __capability uaddr, size_t len)
 {
 	if (linux_remap_address(&uaddr, len)) {
 		if (uaddr == NULL)
 			return (-EFAULT);
-		memcpy(uaddr, kaddr, len);
+		memcpy((void *)(__cheri_addr uintptr_t)uaddr, kaddr, len);
```
Comment (qwattash) [inline sys/compat/linuxkpi/common/src/linux_compat.c:853]: Same as above

File: sys/compat/linuxkpi/common/src/linux_compat.c [hunk 7]
```diff
@@ -824,44 +826,44 @@ linux_remap_address(void **uaddr, size_t len)
...
-linux_clear_user(void *_uaddr, size_t _len)
+linux_clear_user(void * __capability _uaddr, size_t _len)
 {
-	uint8_t *uaddr = _uaddr;
+	uint8_t * __capability uaddr = _uaddr;
 	size_t len = _len;
 
 	/* make sure uaddr is aligned before going into the fast loop */
-	while (((uintptr_t)uaddr & 7) != 0 && len > 7) {
+	while (((__cheri_addr uintptr_t)uaddr & 7) != 0 && len > 7) {
```
Comment (qwattash) [inline sys/compat/linuxkpi/common/src/linux_compat.c:866]: Probably a good place for `__builtin_is_aligned()`. I have an `is_aligned()` macro in the purecap kernel branch that I should probably have merged sooner. We could use the builtin here and then move it to the macro when we merge it.

File: sys/contrib/ipfilter/netinet/fil.c [hunk 8]
```diff
@@ -8009,7 +8024,6 @@ ipf_getnextrule(softc, t, ptr)
 		next = &zero;
 		t->ipt_data = NULL;
 	}
-	it.iri_rule = predict ? next : NULL;
```
Comment (qwattash) [inline sys/contrib/ipfilter/netinet/fil.c:8012]: Why was this line deleted?

File: sys/contrib/ipfilter/netinet/ip_fil.h [hunk 9]
```diff
@@ -1321,7 +1321,7 @@ typedef	int	(*ipftq_delete_fn_t)(struct ipf_main_softc_s *, void *);
 typedef	struct	ipfobj	{
 	u_32_t		ipfo_rev;	/* IPFilter version number */
 	u_32_t		ipfo_size;	/* size of object at ipfo_ptr */
-	void		*ipfo_ptr;	/* pointer to object */
+	void * __kerncap ipfo_ptr;	/* pointer to object */
```
Comment (qwattash) [inline sys/contrib/ipfilter/netinet/ip_fil.h:1324]: nit. keep indentation consistent with the rest of the file
```suggestion
	void			* __kerncap ipfo_ptr;	/* pointer to object */
```

File: sys/dev/mfi/mfi_ioctl.h [hunk 10]
```diff
@@ -57,13 +57,13 @@ union mfi_statrequest {
 #define MAX_SPACE_FOR_SENSE_PTR		32
 union mfi_sense_ptr {
 	uint8_t		sense_ptr_data[MAX_SPACE_FOR_SENSE_PTR];
-	void 		*user_space;
+	void * __kerncap user_space;
```
Comment (qwattash) [inline sys/dev/mfi/mfi_ioctl.h:60]: nit. Keep indentation consistent with rest of file
```suggestion
	void			* __kerncap user_space;
```

File: sys/dev/ti/if_ti.c [hunk 11]
```diff
@@ -592,12 +594,12 @@ ti_copy_mem(struct ti_softc *sc, uint32_t tigon_addr, uint32_t len,
 					    sc->ti_membuf2, segsize,
 					    TI_SWAP_NTOH);
 					TI_UNLOCK(sc);
-					bcopy(&sc->ti_membuf2[segresid], ptr,
+					bcopy(&sc->ti_membuf2[segresid], (void *)(__cheri_addr uintptr_t)ptr,
```
Comment (qwattash) [inline sys/dev/ti/if_ti.c:597]: `(void *)(__cheri_addr uintptr_t)ptr` will strip the tag in the purecap kernel.
```suggestion
					bcopy(&sc->ti_membuf2[segresid], (__cheri_fromcap void *)ptr,
```

File: sys/dev/ti/if_ti.c [hunk 12]
```diff
@@ -592,12 +594,12 @@ ti_copy_mem(struct ti_softc *sc, uint32_t tigon_addr, uint32_t len,
...
 					    segsize - segresid);
 					TI_LOCK(sc);
 					first_pass = 0;
 				} else
-					ti_bcopy_swap(sc->ti_membuf, ptr,
+					ti_bcopy_swap(sc->ti_membuf, (void *)(__cheri_addr uintptr_t)ptr,
```
Comment (qwattash) [inline sys/dev/ti/if_ti.c:602]:
```suggestion
					ti_bcopy_swap(sc->ti_membuf, (__cheri_fromcap void *)ptr,
```

File: sys/dev/ti/if_ti.c [hunk 13]
```diff
@@ -609,7 +611,7 @@ ti_copy_mem(struct ti_softc *sc, uint32_t tigon_addr, uint32_t len,
 				ti_bcopy_swap(sc->ti_membuf2, sc->ti_membuf,
 				    segsize, TI_SWAP_HTON);
 			} else
-				ti_bcopy_swap(ptr, sc->ti_membuf, segsize,
+				ti_bcopy_swap((void *)(__cheri_addr uintptr_t)ptr, sc->ti_membuf, segsize,
```
Comment (qwattash) [inline sys/dev/ti/if_ti.c:614]:
```suggestion
				ti_bcopy_swap((__cheri_fromcap void *)ptr, sc->ti_membuf, segsize,
```

File: sys/dev/ti/if_ti.c [hunk 14]
```diff
@@ -659,7 +661,7 @@ ti_copy_mem(struct ti_softc *sc, uint32_t tigon_addr, uint32_t len,
 				copyout(&tmpval2, ptr, resid);
 				TI_LOCK(sc);
 			} else
-				bcopy(&tmpval2, ptr, resid);
+				bcopy(&tmpval2, (void *)(__cheri_addr uintptr_t)ptr, resid);
```
Comment (qwattash) [inline sys/dev/ti/if_ti.c:664]:
```suggestion
				bcopy(&tmpval2, (__cheri_fromcap void *)ptr, resid);
```

File: sys/dev/ti/if_ti.c [hunk 15]
```diff
@@ -677,7 +679,7 @@ ti_copy_mem(struct ti_softc *sc, uint32_t tigon_addr, uint32_t len,
 				copyin(ptr, &tmpval2, resid);
 				TI_LOCK(sc);
 			} else
-				bcopy(ptr, &tmpval2, resid);
+				bcopy((void *)(__cheri_addr uintptr_t)ptr, &tmpval2, resid);
```
Comment (qwattash) [inline sys/dev/ti/if_ti.c:682]:
```suggestion
				bcopy((__cheri_fromcap void *)ptr, &tmpval2, resid);
```

File: sys/dev/ti/if_ti.c [hunk 16]
```diff
@@ -765,12 +767,12 @@ ti_copy_scratch(struct ti_softc *sc, uint32_t tigon_addr, uint32_t len,
 			if (useraddr)
 				copyout(&tmpval, ptr, 4);
 			else
-				bcopy(&tmpval, ptr, 4);
+				bcopy(&tmpval, (void *)(__cheri_addr uintptr_t)ptr, 4);
```
Comment (qwattash) [inline sys/dev/ti/if_ti.c:770]:
```suggestion
				bcopy(&tmpval, (__cheri_fromcap void *)ptr, 4);
```

File: sys/dev/ti/if_ti.c [hunk 17]
```diff
@@ -765,12 +767,12 @@ ti_copy_scratch(struct ti_softc *sc, uint32_t tigon_addr, uint32_t len,
...
 		} else {
 			if (useraddr)
 				copyin(ptr, &tmpval2, 4);
 			else
-				bcopy(ptr, &tmpval2, 4);
+				bcopy((void *)(__cheri_addr uintptr_t)ptr, &tmpval2, 4);
```
Comment (qwattash) [inline sys/dev/ti/if_ti.c:775]:
```suggestion
				bcopy((__cheri_fromcap void *)ptr, &tmpval2, 4);
```

File: sys/cam/ctl/ctl_frontend_ioctl.c (hunk 2 above)
Comment (bsdjhb) [reply sys/cam/ctl/ctl_frontend_ioctl.c:372]: Userland code that sets `ext_data_ptr` actually sets `ext_data_uptr`.  The union only exists in the kernel, and in purecap userland `ext_data_ptr` is a capability.

File: sys/cam/cam_periph.c (hunk 1 above)
Comment (bsdjhb) [reply sys/cam/cam_periph.c:937]: I don't know, and it could be out-of-tree things like smartctl, etc.  The pass(4) driver used for SCSI pass through commands uses this interface.  camcontrol might be another one.  However, the goal here is not to let freebsd64 versions work, but to let purecap versions work.  Perhaps data_ptrs[] array needs to be an explicit array of capabilities instead.

Ah, yes.  I need to instead go splat a bunch of __kerncap annotations in `union ccb` and rework some of this to assume user capabilities instead.  If we need freebsd64 support we will have to do that separately.

File: sys/cam/scsi/scsi_pass.c (hunk 3 above)
Comment (bsdjhb) [reply sys/cam/scsi/scsi_pass.c:1246]: The problem is you can't pass a capability via var-args in hybrid.  It's not implemented in the ABI for either MIPS or RISC-V.  @jrtc27 had some ideas such as passing a pointer to the capability instead as the ABI to use implicitly, but that needs some toolchain work.

File: sys/cam/scsi/scsi_pass.c (hunk 4 above)
Comment (bsdjhb) [reply sys/cam/scsi/scsi_pass.c:1230]: Here I just kept the existing code, though perhaps I need to adjust this as part of my earlier comment to permit `ds_addr` to be a capability.  I'll know more about how I'd like to handle that (possibly a union around `ds_addr`) when I look at this again.

File: sys/cam/ctl/ctl_frontend_ioctl.c (hunk 2 above)
Comment (qwattash) [reply sys/cam/ctl/ctl_frontend_ioctl.c:372]: Makes sense now, thanks

File: sys/contrib/ipfilter/netinet/fil.c (hunk 8 above)
Comment (bsdjhb) [reply sys/contrib/ipfilter/netinet/fil.c:8012]: It's dead code that isn't used, and it's assigning a kernel pointer (vaddr) to a user pointer (capability) so fails to compile.

File: sys/contrib/ipfilter/netinet/ip_fil.h (hunk 9 above)
Comment (bsdjhb) [reply sys/contrib/ipfilter/netinet/ip_fil.h:1324]: @brooksdavis has previously stated he prefers this version to avoid reindenting the rest of the line as much (e.g. the comment in this case still lines up with the other comments)

File: sys/cam/scsi/scsi_pass.c (hunk 3 above)
Comment (qwattash) [reply sys/cam/scsi/scsi_pass.c:1246]: Right, I forgot that was the reason. I don't have a strong opinion on this, I'd say it depends on how long we plan to support hybrid. It wouldn't reduce the diff too much as we would need to change the format string anyway.

File: sys/compat/linuxkpi/common/src/linux_compat.c (hunk 5 above)
Comment (qwattash) [reply sys/compat/linuxkpi/common/src/linux_compat.c:841]: If it has not changed or has some special meaning with `uintptr_t`, `__cheri_addr` should not be ignored in purecap as I recall explicitly using it some places.

File: sys/compat/linuxkpi/common/src/linux_compat.c (hunk 7 above)
Comment (arichardson) [reply sys/compat/linuxkpi/common/src/linux_compat.c:866]: Do we have wrapper macros for __builtin_is_aligned? I think we changed roundup2 and rounddown2 to use the builtins but not sure about this one.

File: sys/dev/ti/if_ti.c (hunk 11 above)
Comment (arichardson) [reply sys/dev/ti/if_ti.c:597]: Do we really want this to be kernel-ddc relative?

File: sys/dev/ti/if_ti.c (hunk 12 above)
Comment (arichardson) [reply sys/dev/ti/if_ti.c:602]: Same question here. I know in purecap __cheri_fromcap will do the right thing since it just returns the capability unchanged but in hybrid it will be relative to kernel ddc.

File: sys/compat/linuxkpi/common/src/linux_compat.c (hunk 5 above)
Comment (bsdjhb) [reply sys/compat/linuxkpi/common/src/linux_compat.c:841]: But uintptr_t is uintcap_t in purecap, not vaddr_t.  __cheri_addr for a vaddr_t cast makes sense still, but I would expect a uintcap_t cast of a capability pointer to always be a no-op since you aren't losing bits unlike uintptr_t in hybrid (which is vaddr_t) where you are only getting 64 bits as a result of the cast.

File: sys/dev/ti/if_ti.c (hunk 11 above)
Comment (bsdjhb) [reply sys/dev/ti/if_ti.c:597]: Well, making it a `void *` in hybrid is going to be kernel DDC relative anytime it is dereferenced.  Perhaps it should use fromcap?  I'd have to fix the other end of it to use tocap though instead of whatever it does now perhaps to stuff a kernel pointer into the capability to pass into this.

File: sys/net80211/ieee80211_ioctl.c [hunk 18]
```diff
@@ -226,7 +228,7 @@ ieee80211_ioctl_getstastats(struct ieee80211vap *vap, struct ieee80211req *ireq)
 	if (ireq->i_len > sizeof(struct ieee80211req_sta_stats))
 		ireq->i_len = sizeof(struct ieee80211req_sta_stats);
 	/* NB: copy out only the statistics */
-	error = copyout(&ni->ni_stats, (uint8_t *) ireq->i_data + off,
+	error = copyout(&ni->ni_stats, (uint8_t * __capability) ireq->i_data + off,
```
Comment (brooksdavis) [inline sys/net80211/ieee80211_ioctl.c:231]:
```suggestion
	error = copyout(&ni->ni_stats, (uint8_t * __capability)ireq->i_data + off,
```

File: sys/net80211/ieee80211_ioctl.c [hunk 19]
```diff
@@ -492,7 +494,8 @@ getstainfo_common(struct ieee80211vap *vap, struct ieee80211req *ireq,
 		} else
 			get_sta_info(&req, ni);
 		ireq->i_len = space - req.space;
-		error = copyout(p, (uint8_t *) ireq->i_data+off, ireq->i_len);
+		error = copyout(p, (uint8_t * __capability) ireq->i_data + off,
```
Comment (brooksdavis) [inline sys/net80211/ieee80211_ioctl.c:497]:
```suggestion
		error = copyout(p, (uint8_t * __capability)ireq->i_data + off,
```

File: sys/dev/usb/usb_ioctl.h [hunk 20]
```diff
@@ -185,8 +185,8 @@ struct usb_fs_endpoint {
 	uint64_t ppBuffer;
 	uint64_t pLength;
 #else
-	void  **ppBuffer;		/* pointer to userland buffers */
-	uint32_t *pLength;		/* pointer to frame lengths, updated
+	void  * __kerncap * __kerncap ppBuffer;	/* pointer to userland buffers */
+	uint32_t  * __kerncap pLength;	/* pointer to frame lengths, updated
```
Comment (brooksdavis) [inline sys/dev/usb/usb_ioctl.h:189]:
```suggestion
	void * __kerncap * __kerncap ppBuffer;	/* pointer to userland buffers */
	uint32_t * __kerncap pLength;	/* pointer to frame lengths, updated
```

File: sys/net/if_lagg.h [hunk 21]
```diff
@@ -116,7 +116,7 @@ struct lagg_reqall {
 	u_int			ra_proto;		/* lagg protocol */
 
 	size_t			ra_size;		/* size of buffer */
-	struct lagg_reqport	*ra_port;		/* allocated buffer */
+	struct lagg_reqport	* __kerncap ra_port;	/* allocated buffer */
```
Comment (brooksdavis) [inline sys/net/if_lagg.h:119]:
```suggestion
	struct lagg_reqport * __kerncap ra_port;	/* allocated buffer */
```

File: sys/contrib/ipfilter/netinet/ip_fil.h [hunk 22]
```diff
@@ -1443,7 +1443,7 @@ typedef	struct	ipfruleiter {
 	int		iri_active;
 	int		iri_nrules;
 	int		iri_v;		/* No longer used (compatibility) */
-	frentry_t	*iri_rule;
+	frentry_t	* __kerncap iri_rule;
```
Comment (brooksdavis) [inline sys/contrib/ipfilter/netinet/ip_fil.h:1446]:
```suggestion
	frentry_t * __kerncap iri_rule;
```

File: sys/contrib/ipfilter/netinet/ip_fil.h [hunk 23]
```diff
@@ -1458,7 +1458,7 @@ typedef	struct	ipfruleiter {
 typedef	struct	ipfgeniter {
 	int	igi_type;
 	int	igi_nitems;
-	void	*igi_data;
+	void	* __kerncap igi_data;
```
Comment (brooksdavis) [inline sys/contrib/ipfilter/netinet/ip_fil.h:1461]:
```suggestion
	void * __kerncap igi_data;
```

File: sys/contrib/ipfilter/netinet/ip_fil.h [hunk 24]
```diff
@@ -1473,7 +1473,7 @@ typedef	struct	ipfgeniter {
 
 typedef	struct	ipftable {
 	int	ita_type;
-	void	*ita_table;
+	void	* __kerncap ita_table;
```
Comment (brooksdavis) [inline sys/contrib/ipfilter/netinet/ip_fil.h:1476]:
```suggestion
	void * __kerncap ita_table;
```

File: sys/contrib/ipfilter/netinet/ip_lookup.h [hunk 25]
```diff
@@ -39,7 +39,7 @@ typedef	struct	iplookupop	{
 	u_int	iplo_arg;
 	char	iplo_name[FR_GROUPLEN];
 	size_t	iplo_size;	/* sizeof struct at iplo_struct */
-	void	*iplo_struct;
+	void	* __kerncap iplo_struct;
```
Comment (brooksdavis) [inline sys/contrib/ipfilter/netinet/ip_lookup.h:42]:
```suggestion
	void * __kerncap iplo_struct;
```

File: sys/contrib/ipfilter/netinet/ip_lookup.h [hunk 26]
```diff
@@ -84,7 +84,7 @@ typedef	struct	ipflookupiter	{
 	int			ili_nitems;
 	iplookupiterkey_t	ili_lkey;
 	char			ili_name[FR_GROUPLEN];
-	void			*ili_data;
+	void			* __kerncap ili_data;
```
Comment (brooksdavis) [inline sys/contrib/ipfilter/netinet/ip_lookup.h:87]:
```suggestion
	void * __kerncap	ili_data;
```

File: sys/dev/iscsi/iscsi_ioctl.h [hunk 27]
```diff
@@ -220,7 +220,7 @@ struct iscsi_session_remove {
 
 struct iscsi_session_list {
 	unsigned int			isl_nentries;
-	struct iscsi_session_state	*isl_pstates;
+	struct iscsi_session_state	* __kerncap isl_pstates;
```
Comment (brooksdavis) [inline sys/dev/iscsi/iscsi_ioctl.h:223]:
```suggestion
	struct iscsi_session_state * __kerncap isl_pstates;
```

File: sys/dev/iscsi_initiator/iscsi.h [hunk 28]
```diff
@@ -422,11 +422,21 @@ typedef struct {
      union ipdu_u	ipdu;
      u_int		hdr_dig;	// header digest
 
-     ahs_t		*ahs_addr;
+     union {
+#ifdef _KERNEL
+	     ahs_t * __capability ahs_uaddr;
+#endif
+	     ahs_t	*ahs_addr;
+     };
```
Comment (brooksdavis) [inline sys/dev/iscsi_initiator/iscsi.h:430]: I wonder if judicious use of macros could make these sorts of changes smaller without making them too confusing.

File: sys/compat/linuxkpi/common/src/linux_compat.c (hunk 5 above)
Comment (qwattash) [reply sys/compat/linuxkpi/common/src/linux_compat.c:841]: I just checked what the current compiler does and it crashes when casting a pointer to `(__cheri_addr __intcap_t)`, I do agree that it should mirror `__cheri_to/fromcap` and be a no-op in this case, otherwise it is confusing. I'm opening a clang issue about this.

File: sys/compat/linuxkpi/common/src/linux_compat.c (hunk 7 above)
Comment (bsdjhb) [reply sys/compat/linuxkpi/common/src/linux_compat.c:866]: We didn't change roundup/rounddown because they now DTRT as-is.  We had explicitly used the builtin in some places that might have gone back to using round* to minimize diffs with upstream.

**Comment by bsdjhb** [comment]
Ok, I've reworked some bits of this.  The `(__cheri_addr uintptr_t)` casts should generally be replaced with `__cheri_fromcap` and `__cheri_tocap` pointer casts.  I've also marked up `union ccb` so it actually works for purecap and updated CAM to cope.  This meant that the `__USER_CTOPTR` thing is gone now.  I've also added a union around `ds_addr` in `bus_dma_segment_t` to hold pointers since CAM does that.  As a sanity check, the only places that now use `__USER_CAP*` are in freebsd32 and freebsd64 compat, which is the only places that should be using those.

File: sys/compat/linuxkpi/common/src/linux_compat.c (hunk 7 above)
Comment (bsdjhb) [reply sys/compat/linuxkpi/common/src/linux_compat.c]: I did end up using `__builtin_is_aligned` here.

File: sys/cam/cam_periph.c (hunk 1 above)
Comment (bsdjhb) [reply sys/cam/cam_periph.c]: I've reworked `union ccb` to actually be purecap friendly which removed all the `__USER_CAP` uses in this file.

File: sys/cam/scsi/scsi_pass.c (hunk 4 above)
Comment (bsdjhb) [reply sys/cam/scsi/scsi_pass.c]: This no longer uses `__USER_CAP`.

File: sys/compat/linuxkpi/common/src/linux_compat.c (hunk 5 above)
Comment (bsdjhb) [reply sys/compat/linuxkpi/common/src/linux_compat.c]: This now uses `__cheri_fromcap`

File: sys/dev/ti/if_ti.c (hunk 11 above)
Comment (bsdjhb) [reply sys/dev/ti/if_ti.c]: Ended up using `__cheri_fromcap`

File: sys/sys/systm.h [hunk 29]
```diff
@@ -207,10 +207,24 @@ void	kassert_panic(const char *fmt, ...)  __printflike(1, 2);
...
 
+/*
+ * Permit inverting the above.  Assume untagged values fall into the
+ * special case of __USER_CFROMPTR().
+ */
+#define	__USER_CTOPTR(cap, base)					\
+	((void *)(__builtin_cheri_tag_get(cap) ?			\
+	    (uintptr_t)((char * __capability)(cap) -			\
+		(char * __capability)(base)) :				\
+	    (__cheri_addr uintptr_t)(cap)))
```
[the remark this answers is not in the document]
Comment (bsdjhb) [reply sys/sys/systm.h]: I ended up dropping these macros entirely after the `union ccb` changes.

File: sys/dev/iscsi_initiator/iscsi.h (hunk 28 above)
Comment (bsdjhb) [reply sys/dev/iscsi_initiator/iscsi.h:430]: I'm not sure.  In some cases I wonder if I I couldn't get by with making them __kerncap and just fixing some in-kernel uses to use __fromcap and __tocap casts, but usually it seemed like overloaded pointers were used too widely for that.
